### PR TITLE
[Feature] Add employees model, timestamp fixes, add common fields

### PR DESCRIPTION
**Pull request by fivetran-avinash**
## PR Overview
**This PR will address the following Issue/Feature:** [#55]

**This PR will result in the following new package version:** v0.11.0

We changed the data type from timestamp to date for some fields to fix some timezone conversion issues. 

**Please provide the finalized CHANGELOG entry which details the relevant changes included in this PR:**

## Breaking Changes 
- Casted specific timestamp fields across all staging models as dates  where the Netsuite UI does not perform timezone conversion. Keeping these fields as timestamp causes issues in reporting tools that perform automatic timezone conversion.
- As this will change the datatype of the underlying fields, this will require a  `--full-refresh`.

## Feature Updates
- We introduced the `stg_netsuite2__employees` model to bring in data from the `employee` source table. This was brought in to leverage fields like `first_name`, `last_name` and `supervisor` in downstream models in the `dbt_netsuite` transformation package. 
- Adds additional commonly used fields to `accounts`, `subsidiaries`, `transaction_lines`, `transactions`, `transaction_accounting_lines`, `customers`, and `vendors` within the Netsuite2 staging models.

## Under the Hood
- Created new seed data in `integration_tests` to support the new `stg_netsuite2__employees` model, as well as the new fields introduced into the new Netsuite2 staging models.

## Contributors
- [@jmongerlyra](https://github.com/jmongerlyra) ([PR #54](https://github.com/fivetran/dbt_netsuite_source/pull/54))

### Basic Validation
Please acknowledge that you have successfully performed the following commands locally:
- [x] dbt run –full-refresh && dbt test
- [NA] dbt run (if incremental models are present) && dbt test

Before marking this PR as "ready for review" the following have been applied:
- [x] The appropriate issue has been linked, tagged, and properly assigned
- [x] All necessary documentation and version upgrades have been applied
- [x] docs were regenerated (unless this PR does not include any code or yml updates)
- [x] BuildKite integration tests are passing
- [x] Detailed validation steps have been provided below

### Detailed Validation
Please share any and all of your validation steps:

The new staging employees model is successfully being passed through and new data types are being applied for timestamp->date conversions. (validated this with integration_tests data). 

### If you had to summarize this PR in an emoji, which would it be?
 
🩺

### CODE REVIEW

File: models/netsuite2/stg_netsuite2__transactions.sql
Comment (fivetran-joemarkiewicz) [reply models/netsuite2/stg_netsuite2__transactions.sql:34]: When spot checking these transaction fields which weren't changed in this PR in the [Netsuite documentation](https://www.netsuite.com/help/helpcenter/en_US/srbrowser/Browser2021_1/analytics/record/transaction.html) I can see that it looks like `createddate` is listed as a datetime and shouldn't be adjusted. However, `closedate` does look to be a date. Is this how it reflects in the Fivetran synced data? Can we take a closer look at which fields should and shouldn't be converted.

image

**Review by fivetran-joemarkiewicz — CHANGES_REQUESTED** [review]
@fivetran-avinash thanks for coordinating with @jmongerlyra to push these changes forward! 

I was able to perform and initial review and have a few comments and callouts included in my review below. One major note that jumped out at me is to confirm the usage of the `employee` table. I want to make sure we aren't adding a table to be required if we don't see all customers syncing this table. Should we consider adding a variable? I haven't done too much of a deep dive here, but I would like to evaluate if it makes sense to require for all customers or if we should offer more flexibility.

File: CHANGELOG.md
Comment (fivetran-avinash) [reply CHANGELOG.md:6]: Added.

File: CHANGELOG.md
Comment (fivetran-avinash) [reply CHANGELOG.md:10]: Updated.

File: CHANGELOG.md
Comment (fivetran-avinash) [reply CHANGELOG.md:9]: Updated.

File: README.md
Comment (fivetran-avinash) [reply README.md:62]: The usage is on par with several default tables like `accounttype`  and  `accounting_books`. However it's closer to the bottom end of default usage, so we will move to use a variable configuration here.

File: docs/run_results.json
Comment (fivetran-avinash) [reply docs/run_results.json:1]: Removed

File: models/netsuite2/stg_netsuite2__accounts.sql
Comment (fivetran-avinash) [reply models/netsuite2/stg_netsuite2__accounts.sql:32]: If you were to do a query on our internal data, it looks as if `accountsearchdisplayname` is a coalesce of `acctnumber` and `accountsearchdisplaynamecopy`. I presume since `acctnumber` is already present in this model`, the newly converted `accountsearchdisplaynamecopy` to `display_name` field provides the useful information not provided. 

However, `display_name` might be confusing, since technically `accountsearchdisplayname` is supposed to be that. So I wonder if renaming it to `account_name` might be better.

File: models/netsuite2/stg_netsuite2__accounting_periods.sql
Comment (fivetran-avinash) [reply models/netsuite2/stg_netsuite2__accounting_periods.sql:30]: Updated.

File: models/netsuite2/stg_netsuite2__transactions.sql
Comment (fivetran-avinash) [reply models/netsuite2/stg_netsuite2__transactions.sql:33]: Updated.

File: models/netsuite2/stg_netsuite2__transactions.sql
Comment (fivetran-avinash) [reply models/netsuite2/stg_netsuite2__transactions.sql:34]: Good callout. [Based on Jared's PR](https://github.com/fivetran/dbt_netsuite_source/pull/54), it looks as if he was only concerned with columns in the Netsuite UI that weren't performing timezone conversion rather than providing explicit casting of all columns to dates that were originally timestamps. 

I'm not sure if this is outside the scope of this issue though. Should I create another issue to address this or batch this into the exiting one and bring everything in?

(I do wonder if it's best to point out these updates with finance as well, just to see if there are any concerns with casting these fields to dates).

Comment (fivetran-avinash) [reply models/netsuite2/stg_netsuite2__transactions.sql:34]: (`closedate` is itself a timestamp in the Fivetran data.)

File: README.md
Comment (jmongerlyra) [reply README.md:62]: The `employee` object is where user access is managed. I don't think it's possible to have a configuration of NetSuite that doesn't include this table.

File: models/netsuite2/stg_netsuite2__accounts.sql
Comment (jmongerlyra) [reply models/netsuite2/stg_netsuite2__accounts.sql:32]: The value in `accountsearchdisplaynamecopy` is presented in the UI as `name`. However, in the `stg_netsuite2__accounts` model, `fullname` was already aliased to `name`.

@fivetran-avinash You are correct that `accountsearchdisplayname` is a concatenation of `acctnumber` and `accountsearchdisplaynamecopy`. However that value is not in the Netsuite UI. I don't think aliasing `accountsearchdisplaynamecopy` as `display_name` should cause confusion. Calling it `name` would be preferable but that could cause breakage.

File: models/netsuite2/stg_netsuite2__transactions.sql
Comment (jmongerlyra) [reply models/netsuite2/stg_netsuite2__transactions.sql:34]: Below is the reasoning from the PR on casting specific fields as `date`. 

> Keeping these fields as timestamp causes issues in reporting tools that perform automatic timezone conversion.

Example: 
Someone in GMT +2 would see 8/1 when the NetSuite UI shows 7/31 for a value of `2024-07-31 00:00:00.000000000 +00:00`. The is particularly a problem for fields like `ending_at`.

I only CAST fields that were causing problems. However, you are safe to include others like `closedate` where the documentation states `date`. We would not want to include `createddate` or other fields the documentation lists as `datetime`. The NetSuite UI presents timestamp info for those fields, and perform timezone conversion.

Comment (jmongerlyra) [reply models/netsuite2/stg_netsuite2__transactions.sql:34]: > (closedate is itself a timestamp in the Fivetran data.)

It's a timestamp, however, the time is always midnight UTC. This indicates is should be treated as a `date`. NetSuite docs also list it as a `date`.

File: models/netsuite2/stg_netsuite2__accounts.sql
Comment (fivetran-avinash) [reply models/netsuite2/stg_netsuite2__accounts.sql:32]: @jmongerlyra Thanks for looking into this! Weird that they would have a a field called `accountsearchdisplayname` not available in the UI, but if that is the case then `display_name` makes more sense. Couldn't use `name` because we already have a field named `name` as such.

File: README.md
Comment (fivetran-avinash) [reply README.md:62]: Thanks for providing that context @jmongerlyra !

I think the issue we're seeing from our internal data is that customers have the `employee` table available, but do not always choose to sync it as much as the other existing tables. So we want to provide the optionality to disable this model so it's not always being synced. The variable will still be set to true so it will be available by default.

File: models/netsuite2/stg_netsuite2__transactions.sql
Comment (fivetran-avinash) [reply models/netsuite2/stg_netsuite2__transactions.sql:34]: Thanks for the context @jmongerlyra ! 

So is it safe to assume that if we see a field that is in timestamp format, but the timestamp is always in 00:00:00 format for that field, it's safe to treat it as a `date` and cast it as such?

Comment (jmongerlyra) [reply models/netsuite2/stg_netsuite2__transactions.sql:34]: @fivetran-avinash Midnight UTC values combined with ensuring the NetSuite documentation has the field listed as `date` means it should be safe to CAST as `date`.

Comment (fivetran-avinash) [reply models/netsuite2/stg_netsuite2__transactions.sql]: Thanks @jmongerlyra . Based on that advice and reading of the documentation, I've cast the following additional columns as `date`:

- `firstorderdate` with `customers` (`date_first_order_at` in `stg_netsuite2__customers`)
- `closedate` with `transactions` (`closed_at` in `stg_netsuite2__transactions`)

Every other timestamp had non-00:00:00 format in our development data, or documentation indicated it was datetime. 

(Also `hire_date` would have been a candidate in `employees`, but it was not brought into the staging model)

**Review by fivetran-avinash — COMMENTED** [review]
@fivetran-joemarkiewicz Thanks from some help from @jmongerlyra, this PR is ready for re-review

File: CHANGELOG.md
Comment (fivetran-joemarkiewicz) [inline CHANGELOG.md:6]: This really won't _require_ a full refresh for the staging models. However, it would be for the downstream ones. Let's clarify that.

```suggestion
  - As this will change the datatype of the underlying fields, this will require a  `--full-refresh` for downstream incremental models.
```

File: CHANGELOG.md
Comment (fivetran-joemarkiewicz) [inline CHANGELOG.md:22]: Can you link to the README section to show users _how_ to disable if they are interested.

Comment (fivetran-joemarkiewicz) [reply CHANGELOG.md:22]: It actually looks like this wasn't documented in the README. Would you be able to include this in the same section as the other disable/enable variables.

**Review by fivetran-joemarkiewicz — CHANGES_REQUESTED** [review]
@fivetran-avinash thanks for working on these updates and @jmongerlyra for chiming in and sharing your insights! 

Overall these changes look good! I just have a few final notes to address before this should be good to go! Also, be sure to add the employee disable/enable details in the respective downstream [PR #138](https://github.com/fivetran/dbt_netsuite/pull/138).

File: CHANGELOG.md
Comment (fivetran-avinash) [reply CHANGELOG.md:22]: Well that's annoying. Didn't save the file so it didn't commit. It's in now.

File: CHANGELOG.md
Comment (fivetran-avinash) [reply CHANGELOG.md:6]: Updated.

**Review by fivetran-avinash — COMMENTED** [review]
@fivetran-joemarkiewicz This is ready for re-review.

@jmongerlyra I did notice that [it was mentioned in the feature](https://github.com/fivetran/dbt_netsuite_source/issues/55) that employees fields will be used in downstream models eventually. However, [when reviewing the current PR](https://github.com/fivetran/dbt_netsuite/pull/138) and [the other PR](https://github.com/fivetran/dbt_netsuite/pull/136) you've created, I didn't see any references to employees being brought in yet. 

For now, I removed the references from the CHANGELOG, but was curious how would you eventually bring employees fields into the `dbt_netsuite` transformation package? Would love to learn more!

**Comment by jmongerlyra** [comment]
@fivetran-avinash We do the joins in a production layer downstream of these models to the transaction dataset.

**Comment by fivetran-avinash** [comment]
Thanks @jmongerlyra ! So there are no plans to bring the `employees` data into `dbt_netsuite`, just leverage it from the souce, and join that information in off the end models independent of the package?

**Comment by jmongerlyra** [comment]
@fivetran-avinash No immediate plans, but we could since you all have already parameterized the usage.

File: models/netsuite2/src_netsuite2.yml
Comment (fivetran-joemarkiewicz) [inline models/netsuite2/src_netsuite2.yml:296]: Apologies I didn't catch this after the addition of the `netsuite2__using_employees` variable, but we need to add a config here to allow the disablement of the employee source if it isn't being used.

```suggestion
        description: "{{ doc('employee_table') }}"
        config:
          enabled: "{{ var('netsuite2__using_employees', true) }}"
        columns:
```

**Review by fivetran-joemarkiewicz — CHANGES_REQUESTED** [review]
Thanks for making these changes @fivetran-avinash and helping address my questions along the way @jmongerlyra! This PR is looking to be in a good position! I have one last required change before this will be good to merge, but it should be a quick one. 

Additionally, since the employee model is not used at all downstream at the moment I am a bit cautious/curious if it will impact Quickstart at all (that being the manifest believes it's an end model since it has no downstream dependencies). @fivetran-avinash let's confirm this with our internal team if this will be an issue. If we know it won't then this should be good to go! If it will be an issue, then we can likely avoid it by simply having the default be `false` until we decide to use it downstream. Hopefully the addition of the downstream `public_models` config ensures this won't be an issue.

**Review by fivetran-avinash — COMMENTED** [review]
@fivetran-joemarkiewicz Review comments addressed. Based on Maxwell's response hopefully we are good to go!

**Review by fivetran-joemarkiewicz — APPROVED** [review]
LGTM!

File: CHANGELOG.md
Comment (fivetran-avinash) [reply CHANGELOG.md]: Added the default value information.

File: CHANGELOG.md
Comment (fivetran-avinash) [reply CHANGELOG.md]: Committed.

File: CHANGELOG.md
Comment (fivetran-avinash) [reply CHANGELOG.md]: Updated.

**Review by fivetran-avinash — COMMENTED** [review]
@fivetran-jamie Release notes addressed!

**Review by fivetran-jamie — APPROVED** [review]
lgtm!

**Pull request merged**
